### PR TITLE
storage/backend: set InitialMmapSize to 2GB on 32-bit to prevent overflow

### LINKED ISSUE #4700: constant 10737418240 overflows int
I have a [PR](https://github.com/kubernetes/kubernetes/pull/22604) which failed with following message:
```
etcd/storage/backend/backend.go:41: constant 10737418240 overflows int
```

Full log is [here](http://pr-test.k8s.io/22604/kubernetes-pull-build-test-e2e-gce/31894/).

Note that I am perfectly fine while using Go 1.5

### PR DESCRIPTION
**Pull request by luxas**
Fixes #4700
@xiang90 @hongchaodeng

### CODE REVIEW

**Comment by heyitsanthony** [comment]
arm mmap accepts a 10GB mapping length?

File: storage/backend/boltoption_unix.go [hunk 1]
```diff
@@ -30,5 +30,5 @@ import (
 // silently ignore this flag. Please update your kernel to prevent this.
 var boltOpenOptions = &bolt.Options{
 	MmapFlags:       syscall.MAP_POPULATE,
-	InitialMmapSize: InitialMmapSize,
+	InitialMmapSize: int(InitialMmapSize),
```
Comment (xiang90) [inline storage/backend/boltoption_unix.go]: i think this still overflow 32bit? /cc @hongchaodeng

**Comment by luxas** [comment]
Don't believe that. I'm perfectly fine to move `InitialMmapSize` to `boltoption_unix` and `boltoption_arm` respectively, and set the ARM value to something like 1 or 2 GB

File: storage/backend/boltoption_unix.go (hunk 1 above)
Comment (xiang90) [inline storage/backend/boltoption_unix.go]: /cc @luxas https://play.golang.org/p/HDHHzHbxkO

Probably we should convert to max.int32

**Comment by luxas** [comment]
This was just the first way of getting it compiled. Appreciating discussion.

File: storage/backend/boltoption_unix.go (hunk 1 above)
Comment (luxas) [inline storage/backend/boltoption_unix.go]: Yes, probably. See here: https://github.com/boltdb/bolt/blob/master/bolt_arm.go#L4. 
I would prefer moving this to a specific ARM file.

Comment (luxas) [inline storage/backend/boltoption_unix.go]: `2 * 1024 * 1024 * 1024 == math.MaxInt32 + 1`, so `math.MaxInt32` sgtm

Comment (luxas) [inline storage/backend/boltoption_unix.go]: `boltdb` maxMapSize equals `math.MaxInt32` too

**Comment by xiang90** [comment]
@luxas 

> This was just the first way of getting it compiled. Appreciating discussion.

What will be the next steps? We do worry about delivering unreliable software that can subtly break at runtime. (And we were considering about prevent 32bit from building actually)

**Comment by heyitsanthony** [comment]
It'll build, but I am confident it won't work very well:
```
--- FAIL: TestConcurrentApplyAndSnapshotV3 (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x4 pc=0xe8c28]

goroutine 124 [running]:
testing.func·006()
	/home/chz/go1.4/go/src/testing/testing.go:441 +0x188
sync/atomic.storeUint64(0x10d1057c, 0x4000, 0x0)
	/home/chz/go1.4/go/src/sync/atomic/64bit_arm.go:20 +0x48
github.com/coreos/etcd/storage/backend.(*batchTx).commit(0x10e156e0, 0x10e15600)
	/home/chz/golib/src/github.com/coreos/etcd/storage/backend/batch_tx.go:159 +0x394
github.com/coreos/etcd/storage/backend.(*batchTx).Commit(0x10e156e0)
	/home/chz/golib/src/github.com/coreos/etcd/storage/backend/batch_tx.go:114 +0x80
github.com/coreos/etcd/storage/backend.newBatchTx(0x10d10550, 0x10e13540)
	/home/chz/golib/src/github.com/coreos/etcd/storage/backend/batch_tx.go:46 +0x80
github.com/coreos/etcd/storage/backend.newBackend(0x10e0b0b0, 0x28, 0x5f5e100, 0x0, 0x2710, 0x1f)
	/home/chz/golib/src/github.com/coreos/etcd/storage/backend/backend.go:103 +0x288
github.com/coreos/etcd/storage/backend.NewTmpBackend(0x5f5e100, 0x0, 0x2710, 0x773b00, 0x0, 0x0)
	/home/chz/golib/src/github.com/coreos/etcd/storage/backend/backend.go:304 +0x1c0
github.com/coreos/etcd/storage/backend.NewDefaultTmpBackend(0x10e028ec, 0x0, 0x0)
	/home/chz/golib/src/github.com/coreos/etcd/storage/backend/backend.go:308 +0x50
github.com/coreos/etcd/etcdserver.TestConcurrentApplyAndSnapshotV3(0x10df68a0)
	/home/chz/golib/src/github.com/coreos/etcd/etcdserver/server_test.go:936 +0x9e8
testing.tRunner(0x10df68a0, 0xb3f5ac)
	/home/chz/go1.4/go/src/testing/testing.go:447 +0xf4
created by testing.RunTests
	/home/chz/go1.4/go/src/testing/testing.go:555 +0x7e0
```

**Comment by luxas** [comment]
Sorry, git error when updating fork. Fixing...

**Comment by luxas** [comment]
Well, now we have a much better alternative.
If `linux,arm`, set a lower (2GB) map size, 'cause it's 32-bit. If not, do as usual.
This won't mess with `linux,amd64` code at all, so we don't have to be afraid of delivering unreliable software.

Thanks for your patience. Now we have a "real" solution. 
The earlier commit was just for starting the discussion with something that compiled.
WDYT? @xiang90 @heyitsanthony @hongchaodeng

**Comment by xiang90** [comment]
@luxas This only fixes build. We worry more about the runtime issues mentioned here https://github.com/coreos/etcd/issues/358. We are not able to fix all issues in the repo since we have a lot of dependencies.

**Comment by luxas** [comment]
Yeah I know that, but can you merge this when ready and shift the blocker to e.g. 32-bit intel or something instead. 
`arm` 32-bit is already considered hackish, and you have stated in the readme that you doesn't support 32-bit.

**Comment by heyitsanthony** [comment]
@luxas I just tried `GOARCH=386 ./build` and `GOARCH=arm ./build` using master; they both work fine. What does this fix?

**Comment by luxas** [comment]
Yeah, of course `arm` and `386` builds with #4721. They just won't work. 
Try to `bin/etcd` your 32-bit binary and it says `unsupported`.
This fixes the real 32-bit compilation, 'cause certainly folks like me will continue to build `etcd` for 32-bit.

**Comment by luxas** [comment]
@heyitsanthony The patch sets `InitialMmapSize` to `math.MaxInt32`, which is `2GB` for `32-bit` platforms

**Comment by luxas** [comment]
You may easily run `32-bit` integration tests:
```
$ git diff
diff --git a/build b/build
index e08bcc4..ce56aa8 100755
--- a/build
+++ b/build
@@ -29,5 +29,5 @@ else
 fi
 
 # Static compilation is useful when etcd is run in a container
-CGO_ENABLED=0 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "-s -X ${REPO_PATH}/version.GitSHA${LINK_OPERATOR}${GIT_SHA}" -o bin/etcd ${REPO_PATH}
-CGO_ENABLED=0 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "-s" -o bin/etcdctl ${REPO_PATH}/etcdctl
+CGO_ENABLED=0 GOARCH=386 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "-s -X ${REPO_PATH}/version.GitSHA${LINK_OPERATOR}${GIT_SHA}" -o bin/etcd ${REPO_PATH}
+CGO_ENABLED=0 GOARCH=386 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "-s" -o bin/etcdctl ${REPO_PATH}/etcdctl
diff --git a/etcdmain/etcd.go b/etcdmain/etcd.go
index e6cdf64..3a7456a 100644
--- a/etcdmain/etcd.go
+++ b/etcdmain/etcd.go
@@ -13,7 +13,6 @@
 // limitations under the License.
 
 // TODO: support arm64
-// +build amd64
 
 package etcdmain
 
diff --git a/etcdmain/etcd_phony.go b/etcdmain/etcd_phony.go
index b918b42..67ef009 100644
--- a/etcdmain/etcd_phony.go
+++ b/etcdmain/etcd_phony.go
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// +build !amd64
+// +build ignore
 
 package etcdmain
```
With those small changes, it's just `INTEGRATION=y ./test`, and it will test `32-bit`. 
In my tests `32-bit` did score exactly as `64-bit`. I will run perf tests also and fix some alignments I've found

**Comment by heyitsanthony** [comment]
> In my tests 32-bit did score exactly as 64-bit.

Interesting. I've tested it on two separate machines (arm, x86) using both linux and darwin; the tests crash in both cases without #4733 but work fine on amd64. This sort of thing is exactly why 32-bit is considered unstable and not ready for general release.

**Comment by luxas** [comment]
I've run `etcd` 2.2 on my Raspberry Pis for 2-3 weeks non-stop without any problems. Also, many folks have used my project https://github.com/luxas/kubernetes-on-arm with `etcd` without reporting any such issues.

**Comment by luxas** [comment]
@heyitsanthony @xiang90 Updated commit message. ptal

**Comment by heyitsanthony** [comment]
I'm not sure these specializations are necessary for the time being. I'm running tests on 386 and arm now to see if it matters.

**Comment by luxas** [comment]
I think they are. 
You included the first commit of this PR in #4721 (`InitialMmapSize` first to `int64`, then back to `int`), but `InitialMmapSize` is still `10GB`, which I think overflows `int32` e.g. here https://github.com/boltdb/bolt/blob/master/bolt_386.go#L4

**Comment by luxas** [comment]
```go
var test = int64(10 * 1024 * 1024 * 1024);
var test2 = int64((2 * 1024 * 1024 * 1024) - 1); // 2 GB, math.MaxInt32
println(int(test))
println(int(test2))
```
gives on `386`
```
-2147483648 //10GB
2147483647 //2GB
```

**Comment by heyitsanthony** [comment]
@luxas 2147483647 is not 2GB; it is not page-aligned. -2147483648 is 2GB. The 386 tests all passed; about to test arm.

I looked at the boltdb code to see what happens with 2GB-- turns out it amounts to a no-op. However, I also found that boltdb is clearly broken for files >= 2GB on 32-bit. If the db file grows >= 2GB, it won't be able to open the file again.
```go
func (db *DB) mmap(minsz int) error {                                                        
        db.mmaplock.Lock()
        defer db.mmaplock.Unlock()                                                           
        
        info, err := db.file.Stat()                                                          
        if err != nil {
                return fmt.Errorf("mmap stat error: %s", err)                                
        } else if int(info.Size()) < db.pageSize*2 {
                return fmt.Errorf("file size too small")                                     
        }                                                                                    
        
        // Ensure the size is at least the minimum size.                                     
        var size = int(info.Size())                                                          
        if size < minsz {
                size = minsz                                                                 
        }
        size, err = db.mmapSize(size)                                                        
        if err != nil {
                return err                                                                   
        } 
```

**Comment by heyitsanthony** [comment]
Crashes near 1GB of DB data on 386.
```
23:08:42 etcd1 | unexpected fault address 0xb744a040
23:08:42 etcd1 | fatal error: fault
23:08:42 etcd1 | [signal 
23:08:42 etcd1 | 0xb code=0x1 addr=0xb744a040 pc=0x8430066]
23:08:42 etcd1 | goroutine 64 [running]:
23:08:42 etcd1 | runtime.throw(0x8872fc0, 0x5)
23:08:42 etcd1 |        /usr/lib/go/src/runtime/panic.go:527 +0x7f fp=0x43f4be08 sp=0x43f4bdfc
23:08:42 etcd1 | runtime.sigpanic()
23:08:42 etcd1 |        /usr/lib/go/src/runtime/sigpanic_unix.go:27 +0x276 fp=0x43f4be34 sp=0x43f4be08
23:08:42 etcd1 | github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt.(*Tx).rollback(0x1903f100)
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt/tx.go:245 +0x66 fp=0x43f4be54 sp=0x43f4be34
23:08:42 etcd1 | github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt.(*Tx).Commit(0x1903f100, 0x0, 0x0)
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt/tx.go:163 +
23:08:42 etcd1 | 0x1f2 fp=0x43f4bf10 sp=0x43f4be54
23:08:42 etcd1 | github.com/coreos/etcd/storage/backend.(*batchTx).commit(0x190bea40, 0x8a18e00)
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/storage/backend/batch_tx.go:139 +0x63 fp=0x43f4bf54 sp=0x43f4bf10
23:08:42 etcd1 | github.com/coreos/etcd/storage/backend.(*batchTx).Commit(0x190bea40)
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/storage/backend/batch_tx.go:114 +0x5a fp=0x43f4bf64 sp=0x43f4bf54
23:08:42 etcd1 | github.com/coreos/etcd/storage/backend.(*backend).run(0x19084820)
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:179 +0xcf fp=0x43f4bfd0 sp=0x43f4bf64
23:08:42 etcd1 | runtime.goexit()
23:08:42 etcd1 |        /usr/lib/go/src/runtime/asm_386.s:1662 +0x1 fp=0x43f4bfd4 sp=0x43f4bfd0
23:08:42 etcd1 | created by github.com/coreos/etcd/storage/backend.newBackend
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/storage/backend/backend.go:108 +0x267
23:08:42 etcd1 | goroutine 1 [chan receive, 1 minutes]:
23:08:42 etcd1 | github.com/coreos/etcd/etcdmain.Main()
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:194 +0x15ee
23:08:42 etcd1 | main.main()
23:08:42 etcd1 |        /home/anthony/src/etcd/gopath/src/github.com/coreos/etcd/main.go:29 +0x17
```

**Comment by heyitsanthony** [comment]
@luxas, (1 << 31) is a no-op because boltdb is using an int so it defaults to using the file size; (1 << 31) - 1 seems like it should immediately trigger the mmapSize failure, but I couldn't reproduce that case. I assume the crash I'm seeing is because `mmapSize` kicks in somewhere once the db grows to 1GB. 

I'm building with:
`GOARCH=386 ./build`

I start a cluster with:
`goreman start -f V3DemoProcfile`

I generate a workload with:
`./tools/benchmark/benchmark  put --total=4096 --key-space-size=4096 --val-size=1048576`

**Comment by luxas** [comment]
@heyitsanthony I agree that `(1 << 31)` isn't a good choice, as it will overflow `int32` and default to file size. I also think `10 * 1024 * 1024 * 1024` (which we have at `master` now) will overflow `int32` and default to the file size, so I think that it should be changed in some way.

This patch sets `initialMmapSize` to `(1 << 31) - 1` (which doesn't overflow, but instead it'll set the size of the `mmap` to `2GB` directly and maybe that consumes too much memory or something) and that could be a solution, but haven't tested that much yet.

**Comment by heyitsanthony** [comment]
OK. I think this should wait until 32-bit is not crashing on >1GB data sets. Otherwise it is misleading and wastes a gigabyte of address space for nothing.

**Comment by luxas** [comment]
@heyitsanthony Do you think #4742 affected the 32-bit situation?

**Comment by heyitsanthony** [comment]
@luxas, nope looks like a corruption fix and flock stuff.

boltdb would probably need a redesign to handle large data sets on 32-bit. The etcd fix can be as simple as putting a quota on the db size (which would be useful elsewhere too, cf. #2393, #4744).

**Comment by luxas** [comment]
Yeah, read your issue some minutes ago, and it seems resonable. 
Should we set the limit somewhere at `1GB` then?

And I assume this 32-bit boltdb issue doesn't affect `arm64` and `ppc64le` at all, so the only thing there before you may support them is some testing and validation.

**Pull request closed without merging**